doc/source/reference/c-api/array.rst

```diff
@@ -1526,6 +1526,198 @@ For all of these macros *arr* must be an instance of a (subclass of)
+          PyArray_Descr **descriptors;
+      } PyArrayMethod_Context
+
+   .. c:member:: PyObject *caller
+
+      The caller, which is typically the ufunc that called the loop. May be
+      ``NULL``.
+
+   .. c:member:: struct PyArrayMethodObject_tag *method
+
+      The method "self". Currently this object is an opaque pointer.
+
+   .. c:member:: PyArray_Descr **descriptors
+
+      An array of descriptors for the ufunc loop, filled in by
+      ``resolve_descriptors``. The length of the array is ``nin`` + ``nout``.
+
+.. c:type:: PyArrayMethod_Spec
+
+   A struct used to register an ArrayMethod with NumPy. We use the slots
+   mechanism used by the Python limited API. See below for the slot definitions.
+
+   .. code-block:: c
+
+       typedef struct {
+          const char *name;
+          int nin, nout;
+          NPY_CASTING casting;
+          NPY_ARRAYMETHOD_FLAGS flags;
+          PyArray_DTypeMeta **dtypes;
+          PyType_Slot *slots;
+       } PyArrayMethod_Spec;
+
+   .. c:member:: const char *name
+
+      The name of the loop.
+
+   .. c:member:: int nin
+
+      The number of input operands
+
+   .. c:member:: int nout
+
+      The number of output operands.
+
+   .. c:member:: NPY_CASTING casting
+
+      Used to indicate how minimally permissive a casting operation should
+      be. For example, if a cast operation might in some circumstances be safe,
+      but in others unsafe, then ``NPY_UNSAFE_CASTING`` should be set. Not used
+      for ufunc loops but must still be set.
+
+   .. c:member:: NPY_ARRAYMETHOD_FLAGS flags
+
+      The flags set for the method.
+
+   .. c:member:: PyArray_DTypeMeta **dtypes
+
+      The DTypes for the loop. Must be ``nin`` + ``nout`` in length.
+
+   .. c:member:: PyType_Slot *slots
+
+      An array of slots for the method. Slot IDs must be one of the values
+      below.
+
+ArrayMethod Slots
+~~~~~~~~~~~~~~~~~
+
+These are used to identify which kind of function an ArrayMethod slot
+implements.
+
+.. c:macro:: NPY_METH_resolve_descriptors
+
+   The function used to set the descriptors for an operation based on the
+   descriptors of the operands. For example, a ufunc operation with two input
+   operands and one output operand that is called without ``out`` being set
+   in the python API, ``resolve_descriptors`` will be passed the descriptors
+   for the two operands and determine the correct descriptor to use for the
+   output based on the output DType set for the ArrayMethod. If ``out`` is
+   set, then the output descriptor would be passed in as well and should not
+   be overriden.
+
+.. c:macro:: NPY_METH_get_reduction_initial
+
+   A function that gets the initial value to use for a reduction.
+
+.. c:macro:: NPY_METH_strided_loop
+.. c:macro:: NPY_METH_contiguous_loop
+.. c:macro:: NPY_METH_unaligned_strided_loop
+.. c:macro:: NPY_METH_unaligned_contiguous_loop
+.. c:macro:: NPY_METH_contiguous_indexed_loop
+
+   TODO: I need help explaining the differences between these
```

Comment:
Feel free to shorten or rewrite, but a suggestion to get you started:
```suggestion
   One dimensional strided loops implementing the behavior (either a ufunc
   or cast).  In most cases, ``NPY_METH_strided_loop`` is the generic and only
   version that needs to be implemented.
   ``NPY_METH_contiguous_loop`` can be implemented additionally as a more
   light-weight/faster version and it is used when all inputs and outputs
   are contiguous.
   To deal with possibly unaligned data, NumPy needs to be able to copy
   unaligned to aligned data.  When implementing a new DType, the "cast" or
   copy for it needs to implement ``NPY_METH_unaligned_strided_loop ``.
   Unlike the normal versions, this loop must not assume that the data can be
   accessed in an aligned fashion.  These loops must copy each value before
   accessing or storing::
   
       type_in in_value;
       type_out out_value
       memcpy(&value, in_data, sizeof(type_in));
       out_value = in_value;
       memcpy(out_data, &out_value, sizeof(type_out)
   
   while a normal loop can just use `*(type_out *)out_data = *(type_in)in_data;`.
   The unaligned loops are currently only used in casts and will never be picked
   in ufuncs (ufuncs create a temporary copy to ensure aligned inputs).
   These loops are ignored when ``NPY_METH_get_loop`` is defined.
```
Plus, move `NPY_contiguous_index_loop` to later, and just say:  A specialized inner-loop to speed up common `ufunc.at` computations.  (But I would also be happy to leave it out.)

Reply:
p.s. I now see that @ngoldbaum already copied the suggestion - so this is really just to say that I agree!